File: src/main/java/com/expense/tracker/domain/Expense.java
package com.expense.tracker.domain;

import com.expense.tracker.adapter.DecimalUtil;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.util.StringUtils;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDateTime;

import static lombok.AccessLevel.NONE;
import static lombok.AccessLevel.PRIVATE;

/**
 * @author <a href="kuldeepyadav7291@gmail.com">Kuldeep</a>
 */
@NoArgsConstructor(access = PRIVATE)
@Getter
@EqualsAndHashCode(of = "id", callSuper = false)
@Entity
@Table(name = "expense")
public class Expense {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Setter
    private Long id;
    @Column(nullable = false)
    private String category;
    private String description;
    @Column(nullable = false)
    @Getter(NONE)
    private int amountInCents;
    @Column(name = "added_on", nullable = false)
    private LocalDateTime addedOn;

    private Expense(String category, String description, int amount, LocalDateTime addedOn) {
        this.category = category;
        this.description = description;
        this.amountInCents = amount;
        this.addedOn = addedOn;
    }

    public int amount() {
        return amountInCents;
    }

    public BigDecimal amountInDecimal() {
        return DecimalUtil.centsToDecimal(amountInCents);
    }

    public static Expense of(String category, String description, int amount, LocalDateTime addedOn) {
        if(amount <= 0) throw new IllegalArgumentException("Amount cannot be less than or equal to zero");
        return new Expense(
                StringUtils.isEmpty(category) ? "Miscellaneous" : category,
                description,
                amount,
                addedOn
        );
    }
}

File: src/main/java/com/expense/tracker/domain/Expense.kt
package com.expense.tracker.domain

import com.expense.tracker.adapter.DecimalUtil
import java.math.BigDecimal
import java.time.LocalDateTime
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.Table

/**
 * @author [Kuldeep](kuldeepyadav7291@gmail.com)
 */
@Entity
@Table(name = "expense")
class Expense (
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,
        @Column(nullable = false) var category: String,
        val description: String?,
        @Column(nullable = false) val amountInCents: Int,
        @Column(name = "added_on", nullable = false) val addedOn: LocalDateTime
) {

    fun setId(id: Long) {
        this.id = id
    }

    fun amount(): Int {
        return amountInCents
    }

    fun amountInDecimal(): BigDecimal {
        return DecimalUtil.centsToDecimal(amountInCents)
    }

    companion object {
        fun of(category: String?, description: String?, amountInCents: Int, addedOn: LocalDateTime):Expense {
            val validCategory = if(category.isNullOrBlank()) "Miscellaneous" else category
            return Expense(null, validCategory, description, amountInCents, addedOn)
        }
    }

}